Make IPropertySubpattern Internal

333fred opened the pull request:
Updates the structure of IPropertySubpattern to be what we decided in the IOperation design meeting, and makes the implementation internal until we finalize C# 8.
Fixes https://github.com/dotnet/roslyn/issues/33018.

mavasani commented on src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs:
```diff
@@ -57,7 +58,7 @@ public static ValueUsageInfo GetValueUsageInfo(this IOperation operation)
                         return ValueUsageInfo.Write;
 
-                    case IRecursivePatternOperation _:
+                    case IOperation iop when iop.GetType().GetInterfaces().Any(i => i.Name == "IRecursivePatternOperation"):
```
Thanks for updating this. Can you also file a tracking issue to revert this IDE change?

333fred replied:
Added a note to [https://github.com/dotnet/roslyn/issues/33175](https://github.com/dotnet/roslyn/issues/33175).

---
In reply to: [254423185](https://github.com/dotnet/roslyn/pull/33178#discussion_r254423185) [](ancestors = 254423185)

mavasani commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2007,5 +2008,42 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
+                            constantValue: default, isImplicit: isImplicit);
+                    }
+                case null:
+                    return null;
```
Should this also return an invalid operation? At least the doc comment [here](https://github.com/dotnet/roslyn/pull/33178/files#diff-ecc8c3941cade7f6eb4c8c01f88a46c1R16) indicates Member is always non-null. If we do decide to allow null, we should make this explicit in doc comments.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationCloner.cs:
```diff
@@ -549,6 +549,16 @@ public override IOperation VisitRecursivePattern(IRecursivePatternOperation oper
+        internal override IOperation VisitPropertySubpattern(IPropertySubpatternOperation operation, object argument)
+        {
+            return new PropertySubpatternOperation(
+                semanticModel: null,
```
>semanticModel: null, [](start = 16, length = 20)

Semantic model should be preserved by the Cloner.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6547,20 +6549,30 @@ public override IOperation VisitDeclarationPattern(IDeclarationPatternOperation
+                operation.Syntax,
+                IsImplicit(operation),
+                Visit(operation.Member),
+                (IPatternOperation)Visit(operation.Pattern));
```
>(IPatternOperation)Visit(operation.Pattern) [](start = 16, length = 43)

It looks like in this case we are not setting the right substitution for the ```InstanceReferenceKind.PatternInput``` for the case when the ```operation.Pattern``` is a recursive pattern. Given that majority of nodes are internal, I'd suggest to not substitute ```InstanceReferenceKind.PatternInput``` with a capture in this PR and leave it as a regular node. We will have to figure out the right way to deal with this later.

333fred replied:
Added a note to [https://github.com/dotnet/roslyn/issues/33175](https://github.com/dotnet/roslyn/issues/33175) to follow up.

---
In reply to: [254440669](https://github.com/dotnet/roslyn/pull/33178#discussion_r254440669) [](ancestors = 254440669)

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs:
```diff
@@ -739,17 +766,19 @@ void M((int X, int Y)? x, bool b)
-
+                              IPropertySubpatternOperation (OperationKind.None, Type: null) (Syntax: 'Item1: var z')
+                                Member: 
+                                  IFieldReferenceOperation: System.Int32 (System.Int32 X, System.Int32 Y).Item1 (OperationKind.FieldReference, Type: System.Int32, IsImplicit) (Syntax: 'Item1')
```
>IsImplicit [](start = 163, length = 10)

Is this supposed to be implicit?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2007,5 +2008,42 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
+        internal IOperation CreatePropertySubpatternMember(Symbol symbol, SyntaxNode syntax)
+        {
+            var nameSyntax = (syntax is SubpatternSyntax subpatSyntax ? subpatSyntax.NameColon?.Name : null) ?? syntax;
+            bool isImplicit = nameSyntax != syntax;
```
>nameSyntax != syntax [](start = 30, length = 20)

Should this condition be inversed? It looks like we are lacking a test that simply asserts IOperation tree shape for this code path, consider adding one.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2007,5 +2008,42 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
+                    {
+                        var constantValue = field.ConstantValue is null ? default(Optional<object>) : new Optional<object>(field.ConstantValue);
+                        var receiver = new InstanceReferenceOperation(
+                            InstanceReferenceKind.PatternInput, _semanticModel, nameSyntax, field.ContainingType, constantValue, isImplicit: true);
```
>field.ContainingType [](start = 92, length = 20)

It feels like '''field.ContainingType``` is not always the right value to use. The field can come from a base type. I think we would want all PatternInputs within the same pattern to use the same type, matching the ```MatchedType```.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2007,5 +2008,42 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
+                case PropertySymbol property:
+                    {
+                        var receiver = new InstanceReferenceOperation(
+                            InstanceReferenceKind.PatternInput, _semanticModel, nameSyntax, property.ContainingType, constantValue: default, isImplicit: true);
```
>property.ContainingType [](start = 92, length = 23)

Same comment.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2007,5 +2008,42 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
+                        var receiver = new InstanceReferenceOperation(
+                            InstanceReferenceKind.PatternInput, _semanticModel, nameSyntax, property.ContainingType, constantValue: default, isImplicit: true);
+                        return new PropertyReferenceOperation(
+                            property, receiver, ImmutableArray<IArgumentOperation>.Empty, _semanticModel, nameSyntax, property.Type.TypeSymbol,
```
>ImmutableArray<IArgumentOperation>.Empty [](start = 48, length = 40)

In an error case, could we ran into a parameterized property?

333fred replied:
It does not appear so. The closest I could get was an `IConstantPatternOperation` child of the `IPropertyPatternOperation`. Using an indexer like below doesn't parse close to cleanly at all, it results in multiple `IncompleteMember`s on the class.
```C#
class C
{
    (int x, int y) this[int i] { get => default; set {} }
    void M(object o, bool b)
    {
        b = /*<bind>*/o is C { [1]: (var x, var y) }/*</bind>*/;
    }
}
```

---
In reply to: [254445327](https://github.com/dotnet/roslyn/pull/33178#discussion_r254445327) [](ancestors = 254445327)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs:
```diff
@@ -1501,24 +1504,50 @@ public CSharpLazyRecursivePatternOperation(CSharpOperationFactory operationFacto
+            BoundSubpattern subpattern,
+            SyntaxNode syntax,
+            SemanticModel semanticModel)
+            : base(semanticModel, syntax, false)
```
>false [](start = 42, length = 5)

Is this for Implicit? Consider using named parameter.

AlekseyTs commented on src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs:
```diff
@@ -1186,24 +1186,34 @@ public override void VisitRecursivePattern(IRecursivePatternOperation operation)
+
+        internal override void VisitPropertySubpattern(IPropertySubpatternOperation operation)
+        {
+            Assert.True(operation.Member is IMemberReferenceOperation);
```
>Assert.True(operation.Member is IMemberReferenceOperation); [](start = 12, length = 59)

What about error cases?

AlekseyTs commented:
Done with review pass (iteration 2)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [461185324](https://github.com/dotnet/roslyn/pull/33178#issuecomment-461185324) [](ancestors = 461185324)

333fred commented:
@dotnet/roslyn-compiler for a second review.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs:
```diff
@@ -1501,24 +1504,53 @@ public CSharpLazyRecursivePatternOperation(CSharpOperationFactory operationFacto
+        public override ImmutableArray<IPropertySubpatternOperation> CreatePropertySubpatterns()
+        {
+            return _boundRecursivePattern.Properties.IsDefault ? ImmutableArray<IPropertySubpatternOperation>.Empty :
+                _boundRecursivePattern.Properties.SelectAsArray((p, fac) => fac.CreatePropertySubpattern(p, MatchedType), _operationFactory);
```
>MatchedType [](start = 108, length = 11)

Is this going to prevent the lambda from caching the delegate? Consider passing it in a tuple parameter, together with operation factory.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2007,5 +2008,40 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
+                default:
+                    // We should expose the symbol in this case somehow:
+                    // https://github.com/dotnet/roslyn/issues/33175
+                    return OperationFactory.CreateInvalidOperation(_semanticModel, nameSyntax, ImmutableArray<IOperation>.Empty, isImplicit: true);
```
>true [](start = 141, length = 4)

Should we use ```isImplicit``` instead of hard coding ```true```?

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs:
```diff
@@ -882,14 +1002,135 @@ void M((int X, int Y) tuple, bool b)
+      PropertySubpatterns (1):
+          IPropertySubpatternOperation (OperationKind.None, Type: null, IsInvalid) (Syntax: 'NotFound: int x')
+            Member: 
+              IInvalidOperation (OperationKind.Invalid, Type: null, IsInvalid, IsImplicit) (Syntax: 'NotFound')
```
>IsImplicit [](start = 79, length = 10)

It feels like this shouldn't be implicit.

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs:
```diff
@@ -882,14 +1002,135 @@ void M((int X, int Y) tuple, bool b)
+{
+    void M1(object o, bool b)
+    {
+        b = /*<bind>*/o is C1 { Prop[1]: var x }/*</bind>*/;
```
>Prop[1] [](start = 32, length = 7)

Consider adding a test without indexing, just the property name

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs:
```diff
@@ -882,14 +1002,135 @@ void M((int X, int Y) tuple, bool b)
+            Pattern: 
+              IConstantPatternOperation (OperationKind.ConstantPattern, Type: null, IsInvalid) (Syntax: 'Prop[1]') (InputType: ?)
+                Value: 
+                  IInvalidOperation (OperationKind.Invalid, Type: ?, IsInvalid) (Syntax: 'Prop[1]')
```
>'Prop[1]' [](start = 89, length = 9)

Is this the same syntax node as for the IInvalidOperation above?

333fred replied:
The syntax of the top `IPropertySubpatternOperation` and the `IInvalidOperation` that is it's `Member` are the same, they are `SubpatternSyntax` nodes. These lower 2, the `IConstantPatternOperation` and the `IInvalidOperation` you commented on, are `ConstantPatternSyntax` and `ElementAccessExpressionSyntax` nodes, respectively.

---
In reply to: [254478564](https://github.com/dotnet/roslyn/pull/33178#discussion_r254478564) [](ancestors = 254478564)

AlekseyTs commented on src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs:
```diff
@@ -1186,27 +1186,35 @@ public override void VisitRecursivePattern(IRecursivePatternOperation operation)
+        internal override void VisitPropertySubpattern(IPropertySubpatternOperation operation)
+        {
+            Assert.True(operation.Member is IMemberReferenceOperation || operation.Member is IInvalidOperation);
+            var member = (IMemberReferenceOperation)operation.Member;
```
>(IMemberReferenceOperation)operation.Member; [](start = 25, length = 44)

Is this going to throw when ```operation.Member is IInvalidOperation```?

AlekseyTs replied:
Is this not getting caught by tests? Even when IOperation hook is enabled?

---
In reply to: [254480084](https://github.com/dotnet/roslyn/pull/33178#discussion_r254480084) [](ancestors = 254480084)

333fred replied:
I should probably run these tests through the hook, I haven't. But no, this visitor isn't called on our regular IOperation tests.

---
In reply to: [254480510](https://github.com/dotnet/roslyn/pull/33178#discussion_r254480510) [](ancestors = 254480510,254480084)

AlekseyTs commented:
It looks like there are legitimate test failures.

AlekseyTs commented:
Done with review pass (iteration 3)

333fred commented:
There were on the first run, second run hasn't completed yet.

---
In reply to: [461217005](https://github.com/dotnet/roslyn/pull/33178#issuecomment-461217005) [](ancestors = 461217005)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [461217061](https://github.com/dotnet/roslyn/pull/33178#issuecomment-461217061) [](ancestors = 461217061)

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs:
```diff
@@ -882,14 +1002,186 @@ void M((int X, int Y) tuple, bool b)
+
+            var compilation = CreateCompilation(source, new[] { vbCompilation.EmitToImageReference() });
+            compilation.VerifyDiagnostics(
+                // (6,33): error CS0154: The property or indexer 'Prop' cannot be used in this context because it lacks the get accessor
```
>// (6,33): error CS0154: The property or indexer 'Prop' cannot be used in this context because it lacks the get accessor [](start = 16, length = 120)

This error feels misleading, there is a get accessor. Consider opening an issue to follow up.

333fred replied:
Filled https://github.com/dotnet/roslyn/issues/33189.

AlekseyTs commented on src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs:
```diff
@@ -1186,25 +1186,38 @@ public override void VisitRecursivePattern(IRecursivePatternOperation operation)
+            {
+                case IFieldSymbol field:
+                case IPropertySymbol prop:
+                case IErrorTypeSymbol error:
```
>case IErrorTypeSymbol error: [](start = 16, length = 28)

This is probably not a valid case any more.

cston commented on src/Compilers/Core/Portable/Operations/OperationNodes.cs:
```diff
@@ -8089,6 +8088,100 @@ public override ImmutableArray<IPatternOperation> DeconstructionSubpatterns
+            SyntaxNode syntax,
+            bool isImplicit,
+            IOperation member,
+            IPatternOperation pattern)
```
Minor point: It looks like constructors for other `IOperation` implementations have child arguments before `semanticModel, syntax, isImplicit` arguments.

cston commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2007,5 +2008,40 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
+
+        internal IOperation CreatePropertySubpatternMember(Symbol symbol, ITypeSymbol matchedType, SyntaxNode syntax)
+        {
+            var nameSyntax = (syntax is SubpatternSyntax subpatSyntax ? subpatSyntax.NameColon?.Name : null) ?? syntax;
```
>(syntax is SubpatternSyntax subpatSyntax ? subpatSyntax.NameColon?.Name : null) [](start = 29, length = 79)

Could remove the `: null` with `(syntax as SubpatternSyntax)?.NameColon?.Name

cston commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2007,5 +2008,40 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
+            {
+                case FieldSymbol field:
+                    {
+                        var constantValue = field.ConstantValue is null ? default(Optional<object>) : new Optional<object>(field.ConstantValue);
```
>field.ConstantValue [](start = 123, length = 19)

Should this be `field.ContantValue.Value`?

If so, please use the helper method:
```
var constantValue = ConvertToOptional(field.ConstantValue);
```

cston commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2007,5 +2008,40 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
+                case FieldSymbol field:
+                    {
+                        var constantValue = field.ConstantValue is null ? default(Optional<object>) : new Optional<object>(field.ConstantValue);
+                        var receiver = new InstanceReferenceOperation(
```
>InstanceReferenceOperation [](start = 43, length = 26)

Are we testing with static fields and properties?
```
class C
{
    static int F;

    static void M(object o)
    {
        bool b = o is C { F: 1 };
    }
}
```

cston commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:
```diff
@@ -14,7 +14,7 @@ internal sealed partial class CSharpOperationFactory
         private static readonly IConvertibleConversion s_boxedIdentityConversion = Conversion.Identity;
 
-        private static Optional<object> ConvertToOptional(ConstantValue value)
+        internal static Optional<object> ConvertToOptional(ConstantValue value)
```
>internal [](start = 8, length = 8)

Is this change necessary?

333fred commented:
@cston none of these comments seem blocking, so I've added notes to https://github.com/dotnet/roslyn/issues/33175 to follow up.
